<a href="https://colab.research.google.com/github/fralfaro/MAT281/blob/main/docs/labs/lab_03.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# MAT281 - Laboratorio N°03





**Objetivo**: Aplicar técnicas avanzadas de manipulación y análisis de datos con pandas sobre un conjunto real de datos de contenido de Netflix, reforzando buenas prácticas y métodos eficientes sin recurrir a `groupby`, `merge`, `pivot`, ni `join`.



**Dataset**:

Trabajaremos con el archivo `netflix_titles.csv`, que contiene información sobre los títulos disponibles en la plataforma Netflix hasta el año 2021.

| Variable       | Clase     | Descripción                                                                 |
|----------------|-----------|------------------------------------------------------------------------------|
| show_id        | caracter  | Identificador único del título en el catálogo de Netflix.                   |
| type           | caracter  | Tipo de contenido: 'Movie' o 'TV Show'.                                     |
| title          | caracter  | Título del contenido.                                                       |
| director       | caracter  | Nombre del director (puede ser nulo).                                       |
| cast           | caracter  | Lista de actores principales (puede ser nulo).                              |
| country        | caracter  | País o países donde se produjo el contenido.                                |
| date_added     | fecha     | Fecha en la que el título fue agregado al catálogo de Netflix.              |
| release_year   | entero    | Año de lanzamiento original del título.                                     |
| rating         | caracter  | Clasificación por edad (por ejemplo: 'PG-13', 'TV-MA').                      |
| duration       | caracter  | Duración del contenido (minutos o número de temporadas para series).        |
| listed_in      | caracter  | Categorías o géneros en los que está clasificado el contenido.              |
| description    | caracter  | Breve sinopsis del contenido.                                               |




In [32]:
import pandas as pd

# Cargar datos
df = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/netflix_titles.csv')
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...


### Parte 1: Limpieza y preparación

1. Revisar y describir el dataset:

   * ¿Cuántas filas y columnas tiene?
   * ¿Qué tipos de datos hay?
   * ¿Cuántos valores nulos hay por columna?

2. Transformar la columna `date_added` a tipo fecha.

3. Crear columnas auxiliares con `assign`:

   * Año (`year_added`)
   * Mes (`month_added`)



In [2]:
#1 filas y columnas
df.shape

(8807, 12)

Tiene 8807 columnas y 12 filas

In [33]:
#1 tipos de datos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


Hay 2 tipos de datos int64, object

In [34]:
#1 valores nulos
print(df.isnull().sum())

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64


En director:2634, cast:825, country:831, date_added:10, rating:4, duration:3 y en el resto 0 valores nulos

In [35]:
#2
df['date_added'] = pd.to_datetime(df['date_added'], format='mixed', errors='coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       8807 non-null   object        
 1   type          8807 non-null   object        
 2   title         8807 non-null   object        
 3   director      6173 non-null   object        
 4   cast          7982 non-null   object        
 5   country       7976 non-null   object        
 6   date_added    8797 non-null   datetime64[ns]
 7   release_year  8807 non-null   int64         
 8   rating        8803 non-null   object        
 9   duration      8804 non-null   object        
 10  listed_in     8807 non-null   object        
 11  description   8807 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 825.8+ KB


In [36]:
#3
df = df.assign(
    year_added = df['date_added'].dt.year,
    month_added = df['date_added'].dt.month
)
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

  date_added  release_year rating   duration  \
0 2021-09-25          2020  PG-13     90 min   
1 2021-09-24          2021  TV-MA  2 Seasons   
2 2021-09-24          2021  TV-MA   1 Season   
3 2021-09-24          2021  TV-MA   1 Season   
4 2021-09-24          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  year_added  month_added  
0  As her father nears the end of his life, filmm...      2021.0          9.0  
1  After crossing paths at a party, a Cape Town t...      2021.0          9.0  
2  To protect his family from a powerful drug lor...      2021.0          9.0  
3  Feuds, flirtations and toilet talk go down amo...      2021.0          9.0  
4  In a city of coaching centers known to train I...      2021.0          9.0

## Parte 2: Técnicas avanzadas de pandas

4. Utilizar `.loc` para seleccionar películas (`type == 'Movie'`) que fueron agregadas después del año 2018.

5. Utilizar `str.contains()` y `str.extract()`:

   * Filtrar títulos que contienen la palabra 'love' (sin distinguir mayúsculas/minúsculas).
   * Extraer la duración en minutos para las películas desde la columna `duration`.

6. Aplicar `explode()` sobre la columna `listed_in` para obtener una fila por cada género.

7. Obtener un top 10 de géneros más frecuentes utilizando `value_counts()`.

8. Aplicar `where()` y `mask()` para marcar las películas de más de 120 minutos como contenido largo en una nueva columna.

9. Utilizar `.loc` para filtrar películas que cumplen con:

   * Más de 100 minutos de duración.
   * Rating igual a `'R'`.
   * País igual a `'United States'`.

10. Utilizar `.style` para formatear visualmente el top 10 de películas más largas.

In [37]:
#4
df.loc[(df['type'] == 'Movie') & (df['year_added'] > 2018)]

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8798   s8799  Movie                          Zed Plus   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8798        Chandra Prakash Dwivedi   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8798  Adil Hussain, Mona Singh, K.K. Raina, Sanjay M...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8798                                              India 2019-12-31   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8798          2014  TV-MA  131 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8798            Comedies, Dramas, International Movies   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
6    

In [38]:
#5 .contains()
df[df['title'].str.contains('love', case=False, na=False)]

show_id     type                           title  \
25       s26  TV Show            Love on the Spectrum   
158     s159    Movie         Love Don't Cost a Thing   
159     s160    Movie                  Love in a Puff   
206     s207    Movie       LSD: Love, Sex Aur Dhokha   
227     s228    Movie                     Really Love   
...      ...      ...                             ...   
8567   s8568  TV Show  The Year of Happiness and Love   
8575   s8576  TV Show                 This Is My Love   
8649   s8650    Movie           Two Lovers and a Bear   
8721   s8722  TV Show                   What Is Love?   
8725   s8726    Movie            What's Up With Love?   

                       director  \
25                          NaN   
158                   Troy Byer   
159              Pang Ho-cheung   
206            Dibakar Banerjee   
227       Angel Kristi Williams   
...                         ...   
8567                        NaN   
8575                        NaN   
8649                 Kim Nguyen   
8721                        NaN   
8725  Rudy Soedjarwo, Riri Riza   

                                                   cast        country  \
25                                     Brooke Satchwell      Australia   
158   Nick Cannon, Christina Milian, Kenan Thompson,...  United States   
159   Miriam Chin Wah Yeung, Shawn Yue, Singh Hartih...      Hong Kong   
206   Nushrat Bharucha, Anshuman Jha, Neha Chauhan, ...          India   
227   Kofi Siriboe, Yootha Wong-Loi-Sing, Michael Ea...  United States   
...                                                 ...            ...   
8567  Amber Kuo, James Wen, Weber Yang, Sonia Sui, N...         Taiwan   
8575  Jin-mo Joo, Sa-rang Kim, Junior, Ja-in Lee, Su...    South Korea   
8649  Dane DeHaan, Tatiana Maslany, Gordon Pinsent, ...         Canada   
8721  Chris Wu, Jade Chou, Duncan Lai, King Chin, Ki...         Taiwan   
8725  Dian Sastrowardoyo, Nicholas Saputra, Ladya Ch...      Indonesia   

     date_added  release_year rating   duration  \
25   2021-09-21          2021  TV-14  2 Seasons   
158  2021-09-01          2003  PG-13    101 min   
159  2021-09-01          2010  TV-MA    103 min   
206  2021-08-27          2010  TV-MA    112 min   
227  2021-08-25          2020  TV-MA     95 min   
...         ...           ...    ...        ...   
8567 2016-09-01          2009  TV-PG   1 Season   
8575 2017-05-22          2015  TV-14   1 Season   
8649 2017-05-02          2016      R     95 min   
8721 2017-01-01          2012  TV-14   1 Season   
8725 2018-12-08          2002      R    110 min   

                                              listed_in  \
25       Docuseries, International TV Shows, Reality TV   
158                           Comedies, Romantic Movies   
159              Comedies, Dramas, International Movies   
206    Dramas, Independent Movies, International Movies   
227         Dramas, Independent Movies, Romantic Movies   
...                                                 ...   
8567  International TV Shows, Romantic TV Shows, TV ...   
8575  International TV Shows, Korean TV Shows, Roman...   
8649        Dramas, Independent Movies, Romantic Movies   
8721  International TV Shows, Romantic TV Shows, TV ...   
8725      Dramas, International Movies, Romantic Movies   

                                            description  year_added  \
25    Finding love can be hard for anyone. For young...      2021.0   
158   A nerdy teen tries to make himself cool by ass...      2021.0   
159   When the Hong Kong government enacts a ban on ...      2021.0   
206   This provocative drama examines how the voyeur...      2021.0   
227   A rising Black painter tries to break into a c...      2021.0   
...                                                 ...         ...   
8567  Jiang Chen-bo is a loser without prospects for...      2016.0   
8575  A renowned actor who is still pining over the ...      2017.0   
8649  A troubled couple living near the North Pol

In [39]:
#5 .extract()
movies = df[df['type'] == 'Movie'].copy()
movies['duration_minutes'] = movies['duration'].str.extract(r'(\d+)').astype(float)
movies

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
6    

In [40]:
#6
df_exploded = df.assign(listed_in=df['listed_in'].str.split(', ')).explode('listed_in')
df_exploded

show_id     type                 title         director  \
0         s1    Movie  Dick Johnson Is Dead  Kirsten Johnson   
1         s2  TV Show         Blood & Water              NaN   
1         s2  TV Show         Blood & Water              NaN   
1         s2  TV Show         Blood & Water              NaN   
2         s3  TV Show             Ganglands  Julien Leclercq   
...      ...      ...                   ...              ...   
8805   s8806    Movie                  Zoom     Peter Hewitt   
8805   s8806    Movie                  Zoom     Peter Hewitt   
8806   s8807    Movie                Zubaan      Mozez Singh   
8806   s8807    Movie                Zubaan      Mozez Singh   
8806   s8807    Movie                Zubaan      Mozez Singh   

                                                   cast        country  \
0                                                   NaN  United States   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
1     Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2     Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
...                                                 ...            ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...  United States   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...          India   

     date_added  release_year rating   duration                 listed_in  \
0    2021-09-25          2020  PG-13     90 min             Documentaries   
1    2021-09-24          2021  TV-MA  2 Seasons    International TV Shows   
1    2021-09-24          2021  TV-MA  2 Seasons                 TV Dramas   
1    2021-09-24          2021  TV-MA  2 Seasons              TV Mysteries   
2    2021-09-24          2021  TV-MA   1 Season            Crime TV Shows   
...         ...           ...    ...        ...                       ...   
8805 2020-01-11          2006     PG     88 min  Children & Family Movies   
8805 2020-01-11          2006     PG     88 min                  Comedies   
8806 2019-03-02          2015  TV-14    111 min                    Dramas   
8806 2019-03-02          2015  TV-14    111 min      International Movies   
8806 2019-03-02          2015  TV-14    111 min          Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
1     After crossing paths at a party, a Cape Town t...      2021.0   
2     To protect his family from a powerful drug lor...      2021.0   
...                                                 ...         ...   
8805  Dragged from civilian life, a former superhero...      2020.0   
8805  Dragged from civilian life, a former superhero...      2020.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   
8806  A scrappy but poor boy worms his way into a ty...      2019.0   

      month_added  
0             9.0  
1             9.0  
1             9.0  
1             9.0  
2             9.0  
...           ...  
8805          1.0  
8805          1.0  
8806          3.0  
8806          3.0  
8806          3.0  

[19323 rows x 14 columns]

In [41]:
#7
genres = df['listed_in'].str.split(', ').explode()
genres.value_counts().head(10)

listed_in
International Movies        2752
Dramas                      2427
Comedies                    1674
International TV Shows      1351
Documentaries                869
Action & Adventure           859
TV Dramas                    763
Independent Movies           756
Children & Family Movies     641
Romantic Movies              616
Name: count, dtype: int64

In [42]:
#8 where()
movies['long_movie'] = 'No'  # valor por defecto
movies['long_movie'] = movies['long_movie'].where(movies['duration_minutes'] <= 120, 'Yes')
movies

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
6    

In [43]:
#8 mask()
movies['long_movie'] = 'No'  # valor por defecto
movies['long_movie'] = movies['long_movie'].mask(movies['duration_minutes'] > 120, 'Yes')
movies

show_id   type                             title  \
0         s1  Movie              Dick Johnson Is Dead   
6         s7  Movie  My Little Pony: A New Generation   
7         s8  Movie                           Sankofa   
9        s10  Movie                      The Starling   
12       s13  Movie                      Je Suis Karl   
...      ...    ...                               ...   
8801   s8802  Movie                           Zinzana   
8802   s8803  Movie                            Zodiac   
8804   s8805  Movie                        Zombieland   
8805   s8806  Movie                              Zoom   
8806   s8807  Movie                            Zubaan   

                           director  \
0                   Kirsten Johnson   
6     Robert Cullen, José Luis Ucha   
7                      Haile Gerima   
9                    Theodore Melfi   
12              Christian Schwochow   
...                             ...   
8801                Majid Al Ansari   
8802                  David Fincher   
8804                Ruben Fleischer   
8805                   Peter Hewitt   
8806                    Mozez Singh   

                                                   cast  \
0                                                   NaN   
6     Vanessa Hudgens, Kimiko Glenn, James Marsden, ...   
7     Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
9     Melissa McCarthy, Chris O'Dowd, Kevin Kline, T...   
12    Luna Wedler, Jannis Niewöhner, Milan Peschel, ...   
...                                                 ...   
8801  Ali Suliman, Saleh Bakri, Yasa, Ali Al-Jabri, ...   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...   
8804  Jesse Eisenberg, Woody Harrelson, Emma Stone, ...   
8805  Tim Allen, Courteney Cox, Chevy Chase, Kate Ma...   
8806  Vicky Kaushal, Sarah-Jane Dias, Raaghav Chanan...   

                                                country date_added  \
0                                         United States 2021-09-25   
6                                                   NaN 2021-09-24   
7     United States, Ghana, Burkina Faso, United Kin... 2021-09-24   
9                                         United States 2021-09-24   
12                              Germany, Czech Republic 2021-09-23   
...                                                 ...        ...   
8801                       United Arab Emirates, Jordan 2016-03-09   
8802                                      United States 2019-11-20   
8804                                      United States 2019-11-01   
8805                                      United States 2020-01-11   
8806                                              India 2019-03-02   

      release_year rating duration  \
0             2020  PG-13   90 min   
6             2021     PG   91 min   
7             1993  TV-MA  125 min   
9             2021  PG-13  104 min   
12            2021  TV-MA  127 min   
...            ...    ...      ...   
8801          2015  TV-MA   96 min   
8802          2007      R  158 min   
8804          2009      R   88 min   
8805          2006     PG   88 min   
8806          2015  TV-14  111 min   

                                             listed_in  \
0                                        Documentaries   
6                             Children & Family Movies   
7     Dramas, Independent Movies, International Movies   
9                                     Comedies, Dramas   
12                        Dramas, International Movies   
...                                                ...   
8801           Dramas, International Movies, Thrillers   
8802                    Cult Movies, Dramas, Thrillers   
8804                           Comedies, Horror Movies   
8805                Children & Family Movies, Comedies   
8806    Dramas, International Movies, Music & Musicals   

                                            description  year_added  \
0     As her father nears the end of his life, filmm...      2021.0   
6    

In [44]:
#9
filtered_movies = movies.loc[
    (movies['duration_minutes'] > 100) &
    (movies['rating'] == 'R') &
    (movies['country'] == 'United States')
]
filtered_movies

show_id   type                        title               director  \
48       s49  Movie                 Training Day          Antoine Fuqua   
81       s82  Movie                         Kate  Cedric Nicolas-Troyan   
131     s132  Movie  Blade Runner: The Final Cut           Ridley Scott   
139     s140  Movie           Do the Right Thing              Spike Lee   
144     s145  Movie                  House Party        Reginald Hudlin   
...      ...    ...                          ...                    ...   
8678   s8679  Movie              Vincent N Roxxy   Gary Michael Schultz   
8691   s8692  Movie                    Wakefield          Robin Swicord   
8751   s8752  Movie              Wish I Was Here             Zach Braff   
8754   s8755  Movie                       Wolves        Bart Freundlich   
8802   s8803  Movie                       Zodiac          David Fincher   

                                                   cast        country  \
48    Denzel Washington, Ethan Hawke, Scott Glenn, T...  United States   
81    Mary Elizabeth Winstead, Jun Kunimura, Woody H...  United States   
131   Harrison Ford, Rutger Hauer, Sean Young, Edwar...  United States   
139   Danny Aiello, Ossie Davis, Ruby Dee, Richard E...  United States   
144   Christopher Reid, Christopher Martin, Robin Ha...  United States   
...                                                 ...            ...   
8678  Emile Hirsch, Zoë Kravitz, Emory Cohen, Zoey D...  United States   
8691  Bryan Cranston, Jennifer Garner, Jason O'Mara,...  United States   
8751  Zach Braff, Kate Hudson, Donald Faison, Joey K...  United States   
8754  Michael Shannon, Carla Gugino, Taylor John Smi...  United States   
8802  Mark Ruffalo, Jake Gyllenhaal, Robert Downey J...  United States   

     date_added  release_year rating duration  \
48   2021-09-16          2001      R  122 min   
81   2021-09-10          2021      R  106 min   
131  2021-09-01          1982      R  117 min   
139  2021-09-01          1989      R  120 min   
144  2021-09-01          1990      R  104 min   
...         ...           ...    ...      ...   
8678 2017-09-02          2016      R  101 min   
8691 2019-03-02          2016      R  109 min   
8751 2018-08-16          2014      R  106 min   
8754 2019-03-29          2016      R  109 min   
8802 2019-11-20          2007      R  158 min   

                                            listed_in  \
48                                  Dramas, Thrillers   
81                                 Action & Adventure   
131   Action & Adventure, Classic Movies, Cult Movies   
139                  Classic Movies, Comedies, Dramas   
144                             Comedies, Cult Movies   
...                                               ...   
8678                                Dramas, Thrillers   
8691                                           Dramas   
8751             Comedies, Dramas, Independent Movies   
8754        Dramas, Independent Movies, Sports Movies   
8802                   Cult Movies, Dramas, Thrillers   

                                            description  year_added  \
48    A rookie cop with one day to prove himself to ...      2021.0   
81    Slipped a fatal poison on her final job, a rut...      2021.0   
131   In a smog-choked dystopian Los Angeles, blade ...      2021.0   
139   On a sweltering day in Brooklyn, simmering rac...      2021.0   
144   Grounded by his strict father, Kid risks life ...      2021.0   
...                                                 ...         ...   
8678  In rural Louisiana, a terse loner forges a red...      2017.0   
8691  An unhappy father and lawyer quits his suburba...      2019.0   
8751  With his acting career moribund, Aidan Bloom s...      2018.0   
8754  A promising high school basketball player has ...      2019.0   
8802  A political cartoonist, a crime reporter and a...      2019.0   

      month_added  duration_minutes long_movie  
48            9.0             122.0    

In [45]:
#10
top10_longest = movies.sort_values(by='duration_minutes', ascending=False).head(10)
top10_longest.style.background_gradient(subset=['duration_minutes'], cmap='Blues')



### Pregunta Desafío

11. ¿Cuáles son las combinaciones más frecuentes de género y rating en el dataset?
    (Sugerencia: utilizar `value_counts` con `subset=["genre", "rating"]` después de aplicar `explode()`).



### Bonus: Análisis de duplicados y limpieza

12. ¿Existen películas con el mismo nombre (`title`) pero con distinto año de lanzamiento (`release_year`)?
13. ¿Cuántos títulos únicos hay en total en la columna `title`?





In [46]:
#11
genres = df[['listed_in','rating']].copy()
genres['listed_in'] = genres['listed_in'].str.split(', ')
genres = genres.explode('listed_in')
genres.value_counts(subset=['listed_in','rating']).head(10)


listed_in               rating
International Movies    TV-MA     1130
                        TV-14     1065
Dramas                  TV-MA      830
International TV Shows  TV-MA      714
Dramas                  TV-14      693
International TV Shows  TV-14      472
Comedies                TV-14      465
TV Dramas               TV-MA      434
Comedies                TV-MA      431
Dramas                  R          375
Name: count, dtype: int64

La combinación más frecuente es genero:International Movies con rating:TV-MA

In [47]:
#12
duplicates = df[df.duplicated(subset=['title','release_year'], keep=False)]
duplicates

Empty DataFrame
Columns: [show_id, type, title, director, cast, country, date_added, release_year, rating, duration, listed_in, description, year_added, month_added]
Index: []

No hay duplicados mismo nombre pero con distinto año de lanzamiento

In [48]:
#13
unique_titles = df['title'].nunique()
unique_titles

8807

Hay 8807 títulos únicos.